Refactor channel search functionality in `tdm_loader`

EspenEnes opened the pull request:
The previous implementation of the channel search function in the `tdm_loader.py` was refactored. Search now matches channels directly without set generation and updates retrieval with more efficient lookups. The modification updates the testing to reflect the changes in the returned search results as well.

Before the refactor searching throe 22000 channels took 6.5min and after it took 0.08s

Also the test has changed to be similar to group_search when searching for "", all channels should be returned.

espenenesNOV committed:
Refactor channel search functionality in `tdm_loader`

The previous implementation of the channel search function in the `tdm_loader.py` was refactored. Search now matches channels directly without set generation and updates retrieval with more efficient lookups. The modification updates the testing to reflect the changes in the returned search results as well.

Before the refactor searching throe 22000 channels took 6.5min and after it took 0.08s

Also the test has changed to be similar when searching for "", all channels should be returned.

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -228,31 +228,27 @@ def channel_search(self, search_term):
+        channel_group_ids = {v: i for i, v in enumerate(x.get("id") for x in self._xml_chgs)}
+
+        for channel in self._root.findall(".//tdm_channel"):
+            if channel_name := channel.find("name").text:
```
I like the walrus operator. However, It's only supporter for python > 3.8 so I would rather not use it here.
```suggestion
            channel_name = channel.find("name").text
            if channel_name:
```

EspenEnes replied:
yes, this i a good point.

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -228,31 +228,27 @@ def channel_search(self, search_term):
+                channel_id = channel.get("id")
+                group_uri = re.findall(r'id\("(.+?)"\)', channel.find("group").text)
+                group_id = channel_group_ids.get(group_uri[0])
+                channels = channel_groups_cache.get(group_id)
```
I would suggest to use an internal function with `@cache` attribute. It's a bit more pythonic and cleaner to read. It would look something like this

```python
from functools import cache

@cache
def get_channels(group_id):
    group = self._xml_chgs[group_id]
    return {v: i for i, v in enumerate(re.findall(r'id\("(.+?)"\)', group.find("channels").text))}

domna replied:
This line would then change to

```suggestion
                channels = get_channels(group_id)
```

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -228,31 +228,27 @@ def channel_search(self, search_term):
+                group_id = channel_group_ids.get(group_uri[0])
+                channels = channel_groups_cache.get(group_id)
+
+                if not channels:
+                    group = self._xml_chgs[group_id]
+                    channels = {v: i for i, v in enumerate(re.findall(r'id\("(.+?)"\)', group.find("channels").text))}
+                    channel_groups_cache[group_id] = channels
```
With the above suggestion this part can be removed entirely

```suggestion
```

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -228,31 +228,27 @@ def channel_search(self, search_term):
+                    group = self._xml_chgs[group_id]
+                    channels = {v: i for i, v in enumerate(re.findall(r'id\("(.+?)"\)', group.find("channels").text))}
+                    channel_groups_cache[group_id] = channels
+                channel_id = channels.get(channel_id)
```
It's a bit odd to see `channel_id` being redefined w/o being used. So I suggest of getting the id here directly.

```suggestion
                channel_id = channels.get(channel.get("id"))
```

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -228,31 +228,27 @@ def channel_search(self, search_term):
+
+        for channel in self._root.findall(".//tdm_channel"):
+            if channel_name := channel.find("name").text:
+                channel_id = channel.get("id")
```
```suggestion
```

I suggest to get the `id` directly below. This line is then not needed anymore.

domna commented on tdm_loader/tdm_loader.py:
```diff
@@ -228,31 +228,27 @@ def channel_search(self, search_term):
-
-        return ind_chg_ch
+        matched_channels = []
+        channel_groups_cache = {}
```
```suggestion
```

With a caching function this dict is not needed anymore

domna reviewed (changes requested):
Thank you for yet another fix :)

It's looking all good but I suggest to use a caching function. It's a bit more pythonic and easier to understand for people reading the code (and is essentially doing what you did "manually").

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

EspenEnes committed:
Update tdm_loader/tdm_loader.py

Co-authored-by: Florian Dobener <github@schroedingerscat.org>

espenenesNOV committed:
Implement caching and optimize channel search in `tdm_loader`

A cache decorator is introduced for the `get_channels` function to speed up repeated lookups, replacing the previous, dictionary lookup.

EspenEnes commented:
Thanks, I have never used the cache tool before, I knew about it, but never used it.
This was a great suggestion.

But should the function be privat?

domna commented:
Thanks for the changes!

> But should the function be privat?

Yes, I think this would be good. You can also nest it into the `channel_group_search`. That way it's not accessible anywhere outside this function (as private functions are still accessible but by convention shouldn't be used). But naming it `_get_channels` is also fine with me.

espenenesNOV committed:
Rename method get_channels to _get_channels in tdm_loader

The public `get_channels` function is renamed to the private method `_get_channels` in 'tdm_loader.py'. This is due to internal use only, to provide more clarity and prevent unintended external access.

espenenesNOV committed:
Correct typo in method name in tdm_loader

The method name "get_channels" has been corrected to "_get_channels" in the 'tdm_loader.py' file. The change was necessitated by the requirement for the method to be private, hence the underscore, as it's only meant for internal use, preventing unintended external access.

EspenEnes commented:
Should the channel search also use the description of the channel when searching?  Or should that be a different function.

domna commented:
> Should the channel search also use the description of the channel when searching? Or should that be a different function.

I think it's better to have a separate function for it or at least a function parameter which defaults to false.

domna reviewed (approved):
LGTM. Thank you for the contribution :)
